File: Assets/_Scripts/Commons/StateUtils/State.cs
using Styx.Entities;
using Styx.Entities.PlayerModule;
using Styx.Managers;
using UnityEngine;

namespace RageCure.StateUtils
{
    public abstract class State<T> : MonoBehaviour
    {
        protected PlayerId PlayerId;

        public void Start()
        {
            PlayerId = (PlayerId ?? GetComponent<PlayerId>());
        }

        public abstract void Enter(T entity);
        public abstract void Execute(T entity);
        public abstract void FixedExecute(T entity);
        public abstract void Exit(T entity);
    }
}

File: Assembly-FSharp/Assembly-FSharp/States/States/States/States/States/States/States/States/States/States/State.fs
namespace Styx.States

open UnityEngine

[<AbstractClass>]
type State<'a> () =
    inherit MonoBehaviour()
    
    abstract member Enter: 'a -> 'a
    abstract member Execute: 'a -> 'a
    abstract member FixedExecute: 'a -> 'a
    abstract member Exit: 'a -> 'a
    

